### PR TITLE
[SPARK-43516][ML][PYTHON][CONNECT] Base interfaces of sparkML for spark3.5: estimator/transformer/model/evaluator

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

 - Defines basic interfaces of Evaluator / Transformer / Model / Evaluator, these interfaces are designed to support both spark connect and legacy spark mode, and are designed to support train / transform / evaluate over either spark dataframe or pandas dataframe
 - Implement a feature transformer `MaxAbsScaler`, `ScalerScaler`
 - Implement a regressor evaluator `RegressorEvaluator` that supports MSE and R2 metric evaluation
 - Implement a summarizer via pure python code that can summarize array type columns on spark dataframe.
 - Some utility methods.

### Why are the changes needed?

Project: Distributed ML <> spark connect

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
Unit tests.

### CODE REVIEW

File: python/pyspark/mlv2/feature.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        min_max_res = summarize_dataframe(dataset, input_col, ["mean", "std"])
+        mean_values = min_max_res["mean"]
+        std_values = min_max_res["std"]
+
+        model = StandardScalerModel(mean_values, std_values)
+        model._resetUid(self.uid)
+        return self._copyValues(model)
+
+
+class StandardScalerModel(Transformer, HasInputCol, HasOutputCol):
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/feature.py:104]: For simple computation logic like `MaxAbsScalerModel /StandardScaler`, I guess we can implement in two ways:

1. leverage torch utils functions (in this PR);
2. leverage built-in SQL functions (may add a few helper functions if needed), for example

```
In [19]: import pyspark.sql.functions as SF

In [20]: import pyspark.ml.functions as MF

In [21]: bdf.show()
+-----+------+---------+
|label|weight| features|
+-----+------+---------+
|  1.0|   1.0|[0.0,5.0]|
|  0.0|   2.0|[1.0,2.0]|
|  1.0|   3.0|[2.0,1.0]|
|  0.0|   4.0|[3.0,3.0]|
+-----+------+---------+


In [22]: bdf.select(SF.posexplode(MF.vector_to_array(bdf.features)).alias("index", "value")).groupBy("index").agg(SF.avg("value"), SF.stddev_samp("value")).show()
+-----+----------+------------------+
|index|avg(value)|stddev_samp(value)|
+-----+----------+------------------+
|    1|      2.75|1.7078251276599332|
|    0|       1.5|1.2909944487358056|
+-----+----------+------------------+

```

File: python/pyspark/mlv2/tests/test_feature.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+from pyspark.ml.linalg import DenseVector, SparseVector, Vectors
+from pyspark.sql import Row
+from pyspark.testing.utils import QuietTest
+from pyspark.testing.mlutils import check_params, SparkSessionTestCase
+
+from pyspark.mlv2.feature import MaxAbsScaler, StandardScaler
+
+
+class FeatureTests(SparkSessionTestCase):
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/tests/test_feature.py:30]: dose this works on Spark Connect?

Comment (WeichenXu123) [reply python/pyspark/mlv2/tests/test_feature.py:30]: I will add a spark connect side test.

File: python/pyspark/mlv2/feature.py (hunk 1 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/feature.py:104]: Discussed with @zhengruifeng offline:

Because the goal is to support fit both pandas DataFrame and spark dataframe (the 2 cases we hope them sharing most of fit implementation code), and we need to ensure the array operation efficient, so current approach (implementing a summarizer via spark pandas UDF) should be best approach.

File: python/pyspark/mlv2/summarizer.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            if metric == "sum":
+                result["sum"] = self.sum_values.copy()
+            if metric == "mean":
+                result["mean"] = self.sum_values / self.count
+            if metric == "std":
+                if self.count <= 1:
+                    raise ValueError(
+                        "Standard deviation evaluation requires more than one row data."
+                    )
+                result["std"] = np.sqrt(
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/summarizer.py:62]: nit: why not reusing `torchmetrics` here?

**Review by zhengruifeng — APPROVED** [review]
LGTM pending CI

**Review by zhengruifeng — APPROVED** [review]
LGTM pending CI

File: python/pyspark/mlv2/evaluation.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#
+
+from pyspark.mlv2.base import Evaluator
+
+from pyspark.ml.param import Param, Params, TypeConverters
+from pyspark.ml.param.shared import HasLabelCol, HasPredictionCol
+from pyspark.mlv2.util import aggregate_dataframe
+
+import torch
+import torcheval.metrics as torchmetrics
```
Comment (zhengruifeng) [inline python/pyspark/mlv2/evaluation.py:25]: is `torcheval` a new dep?

Comment (WeichenXu123) [reply python/pyspark/mlv2/evaluation.py:25]: Yes. I talked with @mengxr about this and he said we can use it :)

File: python/pyspark/mlv2/summarizer.py (hunk 3 above)
Comment (WeichenXu123) [reply python/pyspark/mlv2/summarizer.py:62]: We need to evaluate std values over feature arrays, I think using summarizer will be more efficient.

**Comment by zhengruifeng** [comment]
I think you may need to add `torchmetrics` to https://github.com/apache/spark/blob/f55fdca10b1d9df2d8126cde07a26f89a75ae1d2/dev/infra/Dockerfile#L74
